### ISSUE TITLE
vkCreateInstance fails if an enabled extension is provided only by an implicit layer

### ISSUE THREAD

**Issue by cdotstout** [KhronosGroup/Vulkan-LoaderAndValidationLayers#2315]
The logic in loader_validate_instance_extensions() appears to be:

check_if_known (assume extension is known);
look for an icd supporting the extension (assume there is none);
search for a layer supporting the extension from the list of enabled layers (assume no enabled layers)

Resulting in error:
Instance extension %s not supported by available ICDs or enabled layers.

So it appears that implicit layers are not checked.

**Comment by lenny-lunarg** [comment]
I don't see any problem in the logic for `loader_validate_instance_extensions`. Your summary of the logic for the function is basically correct, but when it checks the layers for extensions, it gets the layers from the `instance_layers` argument of the function. The layer list is loaded in `vkCreateInstance` (in trampoline.c) with `loader_layer_scan`, which loads both implicit and explicit layers. This means that both implicit and explicit layers should get checked when `loader_validate_instance_extensions` gets called.

Do you have an actual situation where this seems to be creating an error? If you just think from reading the code that it looks wrong, then I personally think it looks right. I'm not aware of any production implicit layer that implements an instance extension, so it's possible that we've missed a bug, but everything looks right to me. If you have a repro case for an error, then I could probably do more.

Keep in mind that since the loader rejects instance extensions it doesn't support, and since implicit layers are enabled outside of the application (and may be disabled by default), it's pretty easy to create an error that was caused by other means.

**Comment by cdotstout** [comment]
Yes, I do have a case that is failing; but it's simply an implicit layer that defines an instance extension.  Can provide more detail if necessary.

But coming back to the code, the loop here:

https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/blob/master/loader/loader.c#L4836

Never executes because pCreateInfo->enabledLayerCount is 0.

**Comment by cdotstout** [comment]
Note this is an extension the loader doesn't know about; so I'm effectively using VK_LOADER_DISABLE_INST_EXT_FILTER to disable the filtering.

**Comment by lenny-lunarg** [comment]
Alright, I think I see the problem you're encountering now. At least very least, I think I have enough info to create a test case where I should be able to reproduce this. I'll know for sure once I've tested this, but I think you're right and I'll need to get the function to check the enabled layer list instead of `ppEnabledLayersNames`.

I'll post here once I have a more definite answer.

**Comment by cdotstout** [comment]
Great, thank you.

**Comment by sixzone11** [comment]
I am also experiencing such a problem. Likewise, if an ICD does not provide an extension to be activated via an implicit layer, the loader determines that it can not be used.

The use of this type of implicit layer is quite useful. If you want to use an extension in a particular application, and no ICD supports that extension, but there is a way to bypass it, it would be possible to provide the extension via an implicit layer for the application to run.

However, as mentioned by @cdotstout , the loader does not properly check the extension of the implicit layer, so it is not possible to use the layer in the same way.

To solve this problem, I wrote the following simple code into https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/blob/8f324b0/loader/loader.c#L4873 (latest loader.c):
``` c
// Not in ICDs, search extension list in implicit layers first
for (uint32_t j = 0; j < instance_layers->count; j++) {
    layer_prop = &instance_layers->list[j];
    if (0 == (layer_prop->type_flags & VK_LAYER_TYPE_FLAG_EXPLICIT_LAYER)) {
        extension_prop = get_extension_property(pCreateInfo->ppEnabledExtensionNames[i], &layer_prop->instance_extension_list);
    }
}

if (extension_prop) {
    continue;
}
```

This works as expected, but it should be more considered because the layers that are activated by environment variables are not considered.

**Comment by lenny-lunarg** [comment]
I'll update where I am on this issue, since it sounds like a couple of people could use it soon. I have a fix I wrote to make the loader recognize extensions provided by implicit layers, but I don't want to submit it without further changes because of what @sixzone11 mentioned:

> This works as expected, but it should be more considered because the layers that are activated by environment variables are not considered.

The real problem is actually more general than not working for implicit layers &mdash; the real problem is that the loader will not recognize extensions from any layer that are not listed by the app during instance creation. This applies to implicit layers, explicit layers enabled via `VK_INSTANCE_LAYERS`, and metalayers. I'm currently working on a more general fix which I hope will solve all of these cases. I'll post here when I have that fix ready.

**Comment by lenny-lunarg** [comment]
This issue should be resolved via PR #2353
